promql/engine.go

```diff
@@ -576,45 +579,137 @@ func (ng *Engine) execEvalStmt(ctx context.Context, query *query, s *parser.Eval
 	return mat, warnings, nil
 }
 
-// subqueryOffsetRange returns the sum of offsets and ranges of all subqueries in the path.
-func (ng *Engine) subqueryOffsetRange(path []parser.Node) (time.Duration, time.Duration) {
+// subqueryTimes returns the sum of offsets and ranges of all subqueries in the path.
+// It also returns the min and max timestamp associated with the subquery.
+// If the @ modifier is used, then the offset is w.r.t. the startTime passed.
+func subqueryTimes(path []parser.Node) (time.Duration, time.Duration, int64, int64) {
 	var (
-		subqOffset time.Duration
-		subqRange  time.Duration
+		subqOffset, subqRange      time.Duration
+		minTimestamp, maxTimestamp int64 = math.MaxInt64, math.MinInt64
 	)
 	for _, node := range path {
 		switch n := node.(type) {
 		case *parser.SubqueryExpr:
-			subqOffset += n.Offset
+			subqOffset += n.OriginalOffset // TODO(codesome): fix this calculation
 			subqRange += n.Range
+			if n.Timestamp != nil {
```

Comment:
Shouldn't a timestamp reset the offset here?

Reply:
It should, I will get to it after fixing other things